wgpu-core/src/device/mod.rs

```diff
@@ -1443,9 +1446,15 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
 
                 // update submission IDs
                 for id in comb.trackers.buffers.used() {
-                    assert!(buffer_guard[id].pending_mapping.is_none());
+                    assert!(match &buffer_guard[id].map_state {
+                        resource::BufferMapState::Referenced {
+                            map_pending: pending_op
+                        } => pending_op.is_none(),
+                        _ => true,
+                    });
                     if !buffer_guard[id].life_guard.use_at(submit_index) {
                         device.temp_suspected.buffers.push(id);
+                        buffer_guard[id].map_state = resource::BufferMapState::Referenced { map_pending: None };
```

Comment:
This needs to be corrected, since buffer_guard[id].map_state is not mutable. But this is where it needs to be updated, right?

Reply:
Perhaps the whole idea to track `Referenced` wasn't need? I don't expect us to mutate anything here. We should only be *checking*, i.e.
```rust
if !buffer_guard[id].life_guard.use_at(submit_index) && !buffer_guard[id].is_mapped() {
  device.temp_suspected.buffers.push(id);
}
```
Also add some code that checks for the need to add the buffer to `temp_suspected.buffers` on unmap().

Reply:
The idea behind tracking if it's currently in use (referenced) is so that you can resolve the future immediately if you know it's safe to so do.

Reply:
>Also add some code that checks for the need to add the buffer to temp_suspected.buffers on unmap().

Mutable borrow of the buffer will be needed to call `unmap_buffer()` and change the map_state to `Unmapped`. Should I proceed with it, by obtaining the buffer_guard with `write()` instead of `read()` method?

Reply:
yes, it's fine to change it to `write()` in `unmap_buffer()`

Reply:
Also what should `BufferMapState` enum finally look like?

Reply:
it depends on the scope you want to cover. @aloucks proposal is to track the GPU referenced state. I think it's simpler not to do it, so the enum would look like I originally suggested:
```rust
enum MapState {
  Waiting, // waiting for GPU to be done before mapping
  Active, // mapped
  Idle, // not mapped
}
```